sympy/polys/matrices/ddm.py

```diff
@@ -92,6 +92,22 @@ def __init__(self, rowslist, shape, domain):
         if not (len(self) == m and all(len(row) == n for row in self)):
             raise DDMBadInputError("Inconsistent row-list/shape")
 
+    def getitem(self, i, j):
+        return self[i][j]
+
+    def extract_slice(self, slice1, slice2):
+        m, n = self.shape
+        ri = range(m)[slice1]
+        ci = range(n)[slice2]
+        slice1 = slice(ri.start, ri.stop, ri.step)
+        slice2 = slice(ci.start, ci.stop, ci.step)
+
+        ddm = [row[slice2] for row in self[slice1]]
+        if not ddm:
+            return DDM(ddm, (0, 0), self.domain)
```

Comment:
A matrix with zero rows does not necessarily have zero columns:
```python
In [120]: M = Matrix(range(9)).reshape(3, 3)

In [121]: M[:0, :]
Out[121]: []

In [122]: M[:0, :].shape
Out[122]: (0, 3)
```

Reply:
Sure, will these changes work for `DDM.extract_slice`?:
```
def extract_slice(self, slice1, slice2):
        m, n = self.shape
        ci = range(n)[slice2]
        ddm = [row[slice2] for row in self[slice1]]
        if not ddm:
            return DDM(ddm, (0, len(ci)), self.domain)

        return DDM(ddm, (len(ddm), len(ddm[0])), self.domain)
```
For example:
```
In [2]: dM = DDM([
   ...:         [ZZ(1), ZZ(2), ZZ(3)],
   ...:         [ZZ(4), ZZ(5), ZZ(6)],
   ...:         [ZZ(7), ZZ(8), ZZ(9)]], (3, 3), ZZ)

In [3]: dM.extract_slice(slice(0), slice(0, 3))
Out[3]: []

In [4]: dM.extract_slice(slice(0), slice(0, 3)).shape
Out[4]: (0, 3)
```
Thanks.

Reply:
Yes, that looks better. Since `ci` will usually not be needed I would compute it in the `if` block.